File: ios/Classes/v2ray/utils/Utilities.swift
import Foundation
import UIKit

/// Utilities 类提供各种实用工具方法。
class Utilities {

    /// 复制文件从输入流到指定的 URL。
    /// - Parameters:
    ///   - src: 输入流源
    ///   - dst: 目标文件 URL
    /// - Throws: 如果复制过程中发生错误，会抛出异常
    static func copyFiles(src: InputStream, dst: URL) throws {
        let out = OutputStream(url: dst, append: false)!
        out.open()  // 打开输出流
        defer { out.close() }  // 确保在方法结束时关闭输出流

        var buf = [UInt8](repeating: 0, count: 1024)  // 缓冲区
        while true {
            let len = src.read(&buf, maxLength: buf.count)  // 从输入流读取数据
            if len <= 0 { break }  // 结束条件：无更多数据
            out.write(buf, maxLength: len)  // 写入输出流
        }
    }

    /// 获取用户资产目录路径。
    /// - Returns: 资产目录的路径
    static func getUserAssetsPath() -> String {
        let fileManager = FileManager.default
        // 获取应用支持目录下的 assets 子目录
        guard let extDir = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first?.appendingPathComponent("assets") else {
            return ""
        }

        // 如果资产目录不存在，则创建
        if !fileManager.fileExists(atPath: extDir.path) {
            do {
                try fileManager.createDirectory(at: extDir, withIntermediateDirectories: true, attributes: nil)
            } catch {
                return ""  // 创建目录失败
            }
        }
        return extDir.path  // 返回资产目录路径
    }

    /// 复制必要的资产文件到用户资产目录。
    /// - Parameter context: 当前的视图控制器上下文
    static func copyAssets() {
        let extFolder = getUserAssetsPath()  // 获取资产目录
        let assetFiles = ["geosite.dat", "geoip.dat"]  // 需要复制的资产文件

        // 遍历资产文件并复制
        for asset in assetFiles {
            if let assetPath = Bundle.main.path(forResource: asset, ofType: nil) {
                do {
                    let src = InputStream(fileAtPath: assetPath)!  // 创建输入流
                    let dst = URL(fileURLWithPath: extFolder).appendingPathComponent(asset)  // 目标路径
                    try copyFiles(src: src, dst: dst)  // 复制文件
                } catch {
                    print("copyAssets failed: \(error)")  // 处理复制失败的错误
                }
            }
        }
    }

    /// 将整数转换为两位数字的字符串格式。
    /// - Parameter value: 要转换的整数值
    /// - Returns: 格式化后的字符串
    static func convertIntToTwoDigit(_ value: Int) -> String {
        return String(format: "%02d", value)  // 使用格式化字符串返回两位数
    }

    /// 解析 V2Ray JSON 配置文件并生成 V2rayConfig 对象。
    /// - Parameters:
    ///   - remark: 备注
    ///   - config: JSON 配置字符串
    ///   - blockedApplication: 被阻止的应用列表
    ///   - bypassSubnets: 绕过的子网列表
    /// - Returns: V2rayConfig 对象，或在失败时返回 nil
    static func parseV2rayJsonFile(remark: String, config: String, blockedApplication: [String], bypassSubnets: [String]) -> V2rayConfig? {
        let v2rayConfig = V2rayConfig()  // 创建 V2rayConfig 实例
        v2rayConfig.REMARK = remark  // 设置备注
        v2rayConfig.BLOCKED_APPS = blockedApplication  // 设置被阻止的应用
        v2rayConfig.BYPASS_SUBNETS = bypassSubnets  // 设置绕过的子网
        v2rayConfig.APPLICATION_ICON = AppConfigs.APPLICATION_ICON  // 设置应用图标
        v2rayConfig.APPLICATION_NAME = AppConfigs.APPLICATION_NAME  // 设置应用名称

        do {
            // 将 JSON 字符串转换为字典
            if let configData = config.data(using: .utf8),
               let configJson = try JSONSerialization.jsonObject(with: configData, options: []) as? [String: Any] {

                // 解析入站配置
                if let inbounds = configJson["inbounds"] as? [[String: Any]] {
                    for inbound in inbounds {
                        if let protocolType = inbound["protocol"] as? String {
                            // 根据协议类型设置相应的端口
                            if protocolType == "socks", let port = inbound["port"] as? Int {
                                v2rayConfig.LOCAL_SOCKS5_PORT = port
                            } else if protocolType == "http", let port = inbound["port"] as? Int {
                                v2rayConfig.LOCAL_HTTP_PORT = port
                            }
                        }
                    }
                }

                // 解析出站配置
                if let outbounds = configJson["outbounds"] as? [[String: Any]],
                   let settings = outbounds[0]["settings"] as? [String: Any],
                   let vnext = settings["vnext"] as? [[String: Any]] {
                    v2rayConfig.CONNECTED_V2RAY_SERVER_ADDRESS = vnext[0]["address"] as? String ?? ""  // 服务器地址
                    v2rayConfig.CONNECTED_V2RAY_SERVER_PORT = vnext[0]["port"] as? String ?? ""  // 服务器端口
                }

                // 移除不必要的字段
                var mutableConfigJson = configJson
                mutableConfigJson.removeValue(forKey: "policy")
                mutableConfigJson.removeValue(forKey: "stats")

                // 如果启用流量和速度统计，添加相应配置
                if AppConfigs.ENABLE_TRAFFIC_AND_SPEED_STATISTICS {
                    let policy: [String: Any] = [
                        "levels": ["8": ["connIdle": 300, "downlinkOnly": 1, "handshake": 4, "uplinkOnly": 1]],
                        "system": ["statsOutboundUplink": true, "statsOutboundDownlink": true]
                    ]
                    mutableConfigJson["policy"] = policy
                    mutableConfigJson["stats"] = [:]
                    v2rayConfig.ENABLE_TRAFFIC_STATISTICS = true  // 启用流量统计
                }

                // 将字典转换回 JSON 数据
                let jsonData = try JSONSerialization.data(withJSONObject: mutableConfigJson, options: [])
                v2rayConfig.V2RAY_FULL_JSON_CONFIG = String(data: jsonData, encoding: .utf8)  // 设置完整的 JSON 配置
            }
        } catch {
            print("parseV2rayJsonFile failed: \(error)")  // 处理解析失败的错误
            return nil  // 返回 nil
        }

        return v2rayConfig  // 返回解析后的配置
    }
}

File: android/src/main/java/com/github/blueboytm/flutter_v2ray/v2ray/utils/Utilities.java
package com.github.blueboytm.flutter_v2ray.v2ray.utils;

import android.content.Context;
import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;

import com.github.blueboytm.flutter_v2ray.v2ray.core.V2rayCoreManager;

public class Utilities {

    // 复制文件方法
    public static void CopyFiles(InputStream src, File dst) throws IOException {
        try (OutputStream out = new FileOutputStream(dst)) {
            byte[] buf = new byte[1024]; // 缓冲区
            int len;
            // 读取源文件并写入目标文件
            while ((len = src.read(buf)) > 0) {
                out.write(buf, 0, len);
            }
        }
    }

    // 获取用户资产路径
    public static String getUserAssetsPath(Context context) {
        File extDir = context.getExternalFilesDir("assets"); // 外部文件目录
        if (extDir == null) {
            return "";
        }
        // 如果目录不存在，返回应用内部目录
        if (!extDir.exists()) {
            return context.getDir("assets", 0).getAbsolutePath();
        } else {
            return extDir.getAbsolutePath();
        }
    }

    // 复制资产文件
    public static void copyAssets(final Context context) {
        String extFolder = getUserAssetsPath(context); // 获取用户资产路径
        try {
            String geo = "geosite.dat,geoip.dat"; // 需要复制的资产文件列表
            for (String assets_obj : context.getAssets().list("")) {
                if (geo.contains(assets_obj)) {
                    // 复制指定的资产文件
                    CopyFiles(context.getAssets().open(assets_obj), new File(extFolder, assets_obj));
                }
            }
        } catch (Exception e) {
            Log.e("Utilities", "copyAssets failed=>", e); // 记录错误
        }
    }

    // 将整数转换为两位数字的字符串
    public static String convertIntToTwoDigit(int value) {
        if (value < 10) return "0" + value; // 小于10时补零
        else return value + ""; // 否则直接返回字符串
    }

    // 解析 V2Ray 配置文件
    public static V2rayConfig parseV2rayJsonFile(final String remark, String config, final ArrayList<String> blockedApplication, final ArrayList<String> bypass_subnets) {
        final V2rayConfig v2rayConfig = new V2rayConfig();
        v2rayConfig.REMARK = remark; // 设置备注
        v2rayConfig.BLOCKED_APPS = blockedApplication; // 设置被阻止的应用
        v2rayConfig.BYPASS_SUBNETS = bypass_subnets; // 设置绕过的子网
        v2rayConfig.APPLICATION_ICON = AppConfigs.APPLICATION_ICON; // 设置应用图标
        v2rayConfig.APPLICATION_NAME = AppConfigs.APPLICATION_NAME; // 设置应用名称

        try {
            JSONObject config_json = new JSONObject(config); // 解析 JSON 配置
            try {
                JSONArray inbounds = config_json.getJSONArray("inbounds"); // 获取 inbound 数组
                for (int i = 0; i < inbounds.length(); i++) {
                    try {
                        // 获取 SOCKS5 端口
                        if (inbounds.getJSONObject(i).getString("protocol").equals("socks")) {
                            v2rayConfig.LOCAL_SOCKS5_PORT = inbounds.getJSONObject(i).getInt("port");
                        }
                    } catch (Exception e) {
                        // 忽略异常
                    }
                    try {
                        // 获取 HTTP 端口
                        if (inbounds.getJSONObject(i).getString("protocol").equals("http")) {
                            v2rayConfig.LOCAL_HTTP_PORT = inbounds.getJSONObject(i).getInt("port");
                        }
                    } catch (Exception e) {
                        // 忽略异常
                    }
                }
            } catch (Exception e) {
                Log.w(V2rayCoreManager.class.getSimpleName(), "startCore warn => can't find inbound port of socks5 or http.");
                return null; // 返回 null 如果未找到端口
            }
            try {
                // 获取 V2Ray 服务器地址和端口
                v2rayConfig.CONNECTED_V2RAY_SERVER_ADDRESS = config_json.getJSONArray("outbounds")
                        .getJSONObject(0).getJSONObject("settings")
                        .getJSONArray("vnext").getJSONObject(0)
                        .getString("address");
                v2rayConfig.CONNECTED_V2RAY_SERVER_PORT = config_json.getJSONArray("outbounds")
                        .getJSONObject(0).getJSONObject("settings")
                        .getJSONArray("vnext").getJSONObject(0)
                        .getString("port");
            } catch (Exception e) {
                v2rayConfig.CONNECTED_V2RAY_SERVER_ADDRESS = config_json.getJSONArray("outbounds")
                        .getJSONObject(0).getJSONObject("settings")
                        .getJSONArray("servers").getJSONObject(0)
                        .getString("address");
                v2rayConfig.CONNECTED_V2RAY_SERVER_PORT = config_json.getJSONArray("outbounds")
                        .getJSONObject(0).getJSONObject("settings")
                        .getJSONArray("servers").getJSONObject(0)
                        .getString("port");
            }
            // 移除不必要的政策和统计信息
            try {
                if (config_json.has("policy")) {
                    config_json.remove("policy");
                }
                if (config_json.has("stats")) {
                    config_json.remove("stats");
                }
            } catch (Exception ignore_error) {
                // 忽略错误
            }
            // 启用流量和速度统计
            if (AppConfigs.ENABLE_TRAFFIC_AND_SPEED_STATICS) {
                try {
                    JSONObject policy = new JSONObject();
                    JSONObject levels = new JSONObject();
                    levels.put("8", new JSONObject()
                            .put("connIdle", 300)
                            .put("downlinkOnly", 1)
                            .put("handshake", 4)
                            .put("uplinkOnly", 1));
                    JSONObject system = new JSONObject()
                            .put("statsOutboundUplink", true)
                            .put("statsOutboundDownlink", true);
                    policy.put("levels", levels);
                    policy.put("system", system);
                    config_json.put("policy", policy);
                    config_json.put("stats", new JSONObject());
                    config = config_json.toString(); // 更新配置
                    v2rayConfig.ENABLE_TRAFFIC_STATICS = true; // 启用流量统计
                } catch (Exception e) {
                    // 忽略错误
                }
            }
        } catch (Exception e) {
            Log.e(Utilities.class.getName(), "parseV2rayJsonFile failed => ", e); // 记录错误
            return null; // 返回 null 如果解析失败
        }
        v2rayConfig.V2RAY_FULL_JSON_CONFIG = config; // 保存完整的 JSON 配置
        return v2rayConfig; // 返回 V2rayConfig 对象
    }
}
